### ISSUE TITLE
Support for multiple WebAssembly runtimes?

### ISSUE THREAD

**Issue by syrusakbary** [bytecodealliance/cargo-wasi#13]
Following #3, I'd like to know if you would be interested in supporting multiple WASI runtimes ([wasmer](https://github.com/wasmerio/wasmer), [WAVM](https://github.com/WAVM/WAVM)) so we don't need to fork the project to make them available for the users :)

cc @AndrewScheidecker

**Comment by alexcrichton** [comment]
There’s a lot of instability in standards such as WASI and Interface Types, and around how WebAssembly runtimes work, so I think for the time being full support for multiple runtimes will be hard to achieve. One of the primary goals of this tool is to make the `wasm32-wasi` experience buttery smooth, and trying to juggle lots of runtimes is something I can't commit to at this time and I think may hurt the experience of `cargo wasi`. This will eventually all settle down of course as standards are finished and such, and at that point it definitely makes sense to ensure that all supported runtimes work well with `cargo wasi`. 

For now, I think it makes sense to at least have basic support, and I think the best way to do that is through an env var specifying which runtime to use. Cargo already has an existing standard for this with the `CARGO_TARGET_WASM32_WASI_RUNNER` env var (via Cargo configuration too), and supporting that here for the time being seems reasonable (documenting that it may not work in the future to account for the above). If you'd like to open such a PR I’d gladly merge that.

I do want to say that the way you’re asking for this seems very aggressive to me. You opened #3 the day after I created this repository, and now you’re threatening a fork and use a demanding tone. I’m happy to hear that you think this is a useful project, but I don’t think it’s appropriate to demand that this project work in any particular way.

**Comment by syrusakbary** [comment]
> There’s a lot of instability in standards such as WASI and Interface Types, and around how WebAssembly runtimes work

I agree, however the spec for each of the phases in WASI is well defined. As such, it's easy to actually test each of the features in each of the runtimes to assure a buttery smooth experience. It might result being non trivial (which I hope is not the case), but it something we can work towards if we are all interested.

> For now, I think it makes sense to at least have basic support, and I think the best way to do that is through an env var specifying which runtime to use. Cargo already has an existing standard for this with the CARGO_TARGET_WASM32_WASI_RUNNER env var (via Cargo configuration too), and supporting that here for the time being seems reasonable (documenting that it may not work in the future to account for the above). If you'd like to open such a PR I’d gladly merge that.

That's something I think we can do!

> You opened #3 the day after I created this repository

I think your work on the Rust ecosystem has been ultra positive, therefore I got excited when I saw you were working on this. I thought it could be the perfect opportunity get Wasmer exposed to more users. Nothing else to read from the issue

> Now you’re threatening a fork and use a demanding tone

If this project is biased against one or other runtime, it will be hard for us (Wasmer) to push for it.
We believe the usefulness of this project will be high, and we would like every runtime user being able to use it without being forced to install a specific runtime (all of this assuming a smooth usage with the integrated runtime, users first).
This can have other very positive side effects, like assuring compatible Command Line Interfaces between runtimes.

If my message read as aggressive that was not my intention. I personally want to avoid forks because it might confuse people on what tool to use, and I think is worth being direct with that.
As laid out, if we want `cargo wasi` to support multiple runtimes a fork might be the only way to make it possible in case you think this is not a useful goal for this project.
Personally, I'd like to contribute back rather than maintaining a fork... but that's a suggestion.

I hope, however, that we can get it done through the `CARGO_TARGET_WASM32_WASI_RUNNER` env var 🤞...I'll follow up with a PR soon!
